File: THREE/Math/Frustum.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace THREE
{
    /// <summary>
    /// @author mrdoob / http://mrdoob.com/
    /// @author alteredq / http://alteredqualia.com/
    /// @author bhouston / http://clara.io
    /// @author tengge / https://github.com/tengge1
    /// </summary>
    public class Frustum
    {
        public Plane[] planes;

        public Frustum(Plane p0 = null, Plane p1 = null, Plane p2 = null, Plane p3 = null, Plane p4 = null, Plane p5 = null)
        {
            this.planes = new Plane[] {
                p0?? new Plane(),
                p1?? new Plane(),
                p2?? new Plane(),
                p3?? new Plane(),
                p4?? new Plane(),
                p5?? new Plane()
            };
        }

        public Frustum Set(Plane p0, Plane p1, Plane p2, Plane p3, Plane p4, Plane p5)
        {
            var planes = this.planes;

            planes[0].Copy(p0);
            planes[1].Copy(p1);
            planes[2].Copy(p2);
            planes[3].Copy(p3);
            planes[4].Copy(p4);
            planes[5].Copy(p5);

            return this;
        }

        public Frustum Clone()
        {
            return new Frustum().Copy(this);
        }

        public Frustum Copy(Frustum frustum)
        {
            var planes = this.planes;

            for (var i = 0; i < 6; i++)
            {
                planes[i].Copy(frustum.planes[i]);
            }

            return this;
        }

        public Frustum SetFromMatrix(Matrix4 m)
        {
            var planes = this.planes;
            var me = m.elements;
            double me0 = me[0], me1 = me[1], me2 = me[2], me3 = me[3];
            double me4 = me[4], me5 = me[5], me6 = me[6], me7 = me[7];
            double me8 = me[8], me9 = me[9], me10 = me[10], me11 = me[11];
            double me12 = me[12], me13 = me[13], me14 = me[14], me15 = me[15];

            planes[0].SetComponents(me3 - me0, me7 - me4, me11 - me8, me15 - me12).Normalize();
            planes[1].SetComponents(me3 + me0, me7 + me4, me11 + me8, me15 + me12).Normalize();
            planes[2].SetComponents(me3 + me1, me7 + me5, me11 + me9, me15 + me13).Normalize();
            planes[3].SetComponents(me3 - me1, me7 - me5, me11 - me9, me15 - me13).Normalize();
            planes[4].SetComponents(me3 - me2, me7 - me6, me11 - me10, me15 - me14).Normalize();
            planes[5].SetComponents(me3 + me2, me7 + me6, me11 + me10, me15 + me14).Normalize();

            return this;
        }

        public bool IntersectsSphere(Sphere sphere)
        {
            var planes = this.planes;
            var center = sphere.center;
            var negRadius = -sphere.radius;

            for (var i = 0; i < 6; i++)
            {

                var distance = planes[i].DistanceToPoint(center);

                if (distance < negRadius)
                {
                    return false;
                }
            }

            return true;
        }

        public bool IntersectsBox(Box3 box)
        {
            var p = new Vector3();

            var planes = this.planes;

            for (var i = 0; i < 6; i++)
            {
                var plane = planes[i];

                // corner at max distance

                p.x = plane.normal.x > 0 ? box.max.x : box.min.x;
                p.y = plane.normal.y > 0 ? box.max.y : box.min.y;
                p.z = plane.normal.z > 0 ? box.max.z : box.min.z;

                if (plane.DistanceToPoint(p) < 0)
                {
                    return false;
                }
            }

            return true;
        }

        public bool ContainsPoint(Vector3 point)
        {
            var planes = this.planes;

            for (var i = 0; i < 6; i++)
            {
                if (planes[i].DistanceToPoint(point) < 0)
                {
                    return false;
                }
            }

            return true;
        }
    }
}

File: server/three/frustum.go
// Copyright 2017-2020 The ShadowEditor Authors. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.
//
// For more information, please visit: https://github.com/tengge1/ShadowEditor
// You can also visit: https://gitee.com/tengge1/ShadowEditor
//
// This package is translated from three.js, visit `https://github.com/mrdoob/three.js`
// for more information.

package three

var _sphere = Sphere{}
var _vectorFrustum = Vector3{}

// NewFrustum :
func NewFrustum(p0, p1, p2, p3, p4, p5 Plane) *Frustum {
	return &Frustum{
		[6]Plane{p0, p1, p2, p3, p4, p5},
	}
}

// Frustum :
type Frustum struct {
	Planes [6]Plane
}

// Set :
func (f Frustum) Set(p0, p1, p2, p3, p4, p5 Plane) *Frustum {
	planes := f.Planes

	planes[0].Copy(p0)
	planes[1].Copy(p1)
	planes[2].Copy(p2)
	planes[3].Copy(p3)
	planes[4].Copy(p4)
	planes[5].Copy(p5)

	return &f
}

// Clone :
func (f Frustum) Clone() *Frustum {
	return NewFrustum(f.Planes[0], f.Planes[1], f.Planes[2], f.Planes[3], f.Planes[4], f.Planes[5]).Copy(f)
}

// Copy :
func (f Frustum) Copy(frustum Frustum) *Frustum {
	planes := f.Planes

	for i := 0; i < 6; i++ {
		planes[i].Copy(frustum.Planes[i])
	}

	return &f
}

// SetFromProjectionMatrix :
func (f Frustum) SetFromProjectionMatrix(m Matrix4) *Frustum {
	planes := f.Planes
	me := m.Elements
	me0, me1, me2, me3 := me[0], me[1], me[2], me[3]
	me4, me5, me6, me7 := me[4], me[5], me[6], me[7]
	me8, me9, me10, me11 := me[8], me[9], me[10], me[11]
	me12, me13, me14, me15 := me[12], me[13], me[14], me[15]

	planes[0].SetComponents(me3-me0, me7-me4, me11-me8, me15-me12).Normalize()
	planes[1].SetComponents(me3+me0, me7+me4, me11+me8, me15+me12).Normalize()
	planes[2].SetComponents(me3+me1, me7+me5, me11+me9, me15+me13).Normalize()
	planes[3].SetComponents(me3-me1, me7-me5, me11-me9, me15-me13).Normalize()
	planes[4].SetComponents(me3-me2, me7-me6, me11-me10, me15-me14).Normalize()
	planes[5].SetComponents(me3+me2, me7+me6, me11+me10, me15+me14).Normalize()

	return &f
}

// IntersectsSphere :
func (f Frustum) IntersectsSphere(sphere Sphere) bool {
	planes := f.Planes
	center := sphere.Center
	negRadius := -sphere.Radius

	for i := 0; i < 6; i++ {
		distance := planes[i].DistanceToPoint(center)
		if distance < negRadius {
			return false
		}
	}

	return true
}

// IntersectsBox :
func (f Frustum) IntersectsBox(box Box3) bool {
	var planes = f.Planes
	for i := 0; i < 6; i++ {
		var plane = planes[i]
		// corner at max distance
		if plane.Normal.X > 0 {
			_vectorFrustum.X = box.Max.X
		} else {
			_vectorFrustum.X = box.Min.X
		}
		if plane.Normal.Y > 0 {
			_vectorFrustum.Y = box.Max.Y
		} else {
			_vectorFrustum.Y = box.Min.Y
		}
		if plane.Normal.Z > 0 {
			_vectorFrustum.Z = box.Max.Z
		} else {
			_vectorFrustum.Z = box.Min.Z
		}
		if plane.DistanceToPoint(_vectorFrustum) < 0 {
			return false
		}
	}
	return true
}

// ContainsPoint :
func (f Frustum) ContainsPoint(point Vector3) bool {
	var planes = f.Planes
	for i := 0; i < 6; i++ {
		if planes[i].DistanceToPoint(point) < 0 {
			return false
		}
	}
	return true
}
